src/common/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp

```diff
@@ -51,31 +52,45 @@ ngraph::pass::BatchNormDecomposition::BatchNormDecomposition() {
 
         const auto& input_type = m_input.get_element_type();
         // scale_add = variance + eps
-        auto scale_add = make_shared<opset5::Add>(m_var, opset5::Constant::create(input_type, Shape{}, {eps}));
+        auto scale_add = std::make_shared<opset5::Add>(m_var, opset5::Constant::create(input_type, Shape{}, {eps}));
         // scale = sqrt(variance + eps)
-        auto scale = make_shared<opset5::Sqrt>(scale_add);
+        auto scale = std::make_shared<opset5::Sqrt>(scale_add);
         // Divide `gamma` by `sqrt(variance + eps)`
         auto gamma_div_scale = std::make_shared<opset5::Divide>(m_gamma, scale);
 
         int64_t dims_to_add = m_input.get_partial_shape().rank().get_length() - 2;
         const auto one = op::Constant::create(element::i64, Shape{1}, {1});
         const auto tail_shape_rank = op::Constant::create(element::i64, Shape{1}, {dims_to_add});
         const auto tail_shape = std::make_shared<opset5::Broadcast>(one, tail_shape_rank);
-        const auto C_dim = std::make_shared<opset5::ShapeOf>(m_gamma);
+        // WA: it's necessary to fold m_gamma shape to the constant
+        // to avoid using m_gamma output twice as divide input and shapeOf input
+        // because this will make it impossible to get constant from source
+        // (lower/upper bounds will be invalidated after the first pass through the m_gamma input)
+        const auto C_dim = ngraph::op::util::make_try_fold<opset5::ShapeOf>(m_gamma);
         // create new shape [1, C, 1, 1, ...]
         const auto new_shape = std::make_shared<opset5::Concat>(
             OutputVector{one, C_dim, tail_shape}, 0);
 
-        auto gamma_div_scale_aligned = make_shared<opset5::Reshape>(gamma_div_scale, new_shape, true);
-        auto beta_aligned = make_shared<opset5::Reshape>(m_beta, new_shape, true);
-        auto mean_aligned = make_shared<opset5::Reshape>(m_mean, new_shape, true);
-
-        // input_sub_mean = input - mean
-        auto input_sub_mean = register_new_node<opset5::Subtract>(m_input, mean_aligned);
+        std::shared_ptr<Node> gamma_div_scale_aligned = std::make_shared<opset5::Reshape>(gamma_div_scale, new_shape, true);
+        std::shared_ptr<Node> beta_aligned = std::make_shared<opset5::Reshape>(m_beta, new_shape, true);
+        std::shared_ptr<Node> mean_aligned = std::make_shared<opset5::Reshape>(m_mean, new_shape, true);
+        std::shared_ptr<Node> mean_negative = std::make_shared<opset5::Multiply>(
+            mean_aligned,
+            opset5::Constant::create(mean_aligned->get_output_element_type(0), Shape{}, {-1}));
+
+         if (auto constant = ov::get_constant_from_source(beta_aligned))
+             beta_aligned = constant;
+         if (auto constant = ov::get_constant_from_source(mean_negative))
+             mean_negative = constant;
+         if (auto constant = ov::get_constant_from_source(gamma_div_scale_aligned))
+             gamma_div_scale_aligned = constant;
+
```

Comment:
Please add tests to cover new behaviour

Reply:
Done, thanks! Also, matcher for opset1 was updated to handle dynamic shapes